# Cleaning process

    1. Import the document

In [3092]:
import pandas as pd
import os
import numpy as np
import re
from collections import Counter

In [3093]:
csv_file = "../data/attacks.csv"
sharks = pd.read_csv(csv_file, encoding = "ISO-8859-1")
sharks

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ...         Species   \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
25718              NaN  NaN  ...              NaN   
25719              NaN  NaN  ...              NaN   
25720              NaN  NaN  ...              NaN   
25721              NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
25718                             NaN                       NaN   
25719                             NaN                       NaN   
25720                             NaN                       NaN   
25721                             NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
25718                                                NaN   
25719                                                NaN   
25720                                                NaN   
25721                                                NaN   
25722                                                NaN   

                                                    href Case Number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

    2. Drop the rows and columns with only NaN

    - Rows

In [3094]:
sharks_ = sharks.dropna(how= "all")
sharks_

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
8698            0          NaN     NaN         NaN        NaN   
8699            0          NaN     NaN         NaN        NaN   
8700            0          NaN     NaN         NaN        NaN   
8701            0          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
8698               NaN                             NaN          NaN   
8699               NaN                             NaN          NaN   
8700               NaN                             NaN          NaN   
8701               NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ...         Species   \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
8698               NaN  NaN  ...              NaN   
8699               NaN  NaN  ...              NaN   
8700               NaN  NaN  ...              NaN   
8701               NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
8698                              NaN                       NaN   
8699                              NaN                       NaN   
8700                              NaN                       NaN   
8701                              NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
8698                                                 NaN   
8699                                                 NaN   
8700                                                 NaN   
8701                                                 NaN   
25722                                                NaN   

                                                    href Case Number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

    '' We can see that there are some rows with the case number "0" and "xx" so we need to be sure that apart from the case numbers all the other valuers are NaN, before dropping them. And also be sure that there isn't any row whith case number value but with all the other values as NaN.''

In [3095]:
nan_raws = sharks_[sharks_.drop("Case Number", axis = 1).isna().all(axis= 1)]
nan_raws

Case Number Date  Year Type Country Area Location Activity Name Sex   \
6309            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6310            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6311            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6312            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6313            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
...           ...  ...   ...  ...     ...  ...      ...      ...  ...  ...   
8698            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8699            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8700            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8701            0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
6309   ...      NaN                    NaN  NaN          NaN  NaN   
6310   ...      NaN                    NaN  NaN          NaN  NaN   
6311   ...      NaN                    NaN  NaN          NaN  NaN   
6312   ...      NaN                    NaN  NaN          NaN  NaN   
6313   ...      NaN                    NaN  NaN          NaN  NaN   
...    ...      ...                    ...  ...          ...  ...   
8698   ...      NaN                    NaN  NaN          NaN  NaN   
8699   ...      NaN                    NaN  NaN          NaN  NaN   
8700   ...      NaN                    NaN  NaN          NaN  NaN   
8701   ...      NaN                    NaN  NaN          NaN  NaN   
25722  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
6309            NaN           NaN            NaN         NaN         NaN  
6310            NaN           NaN            NaN         NaN         NaN  
6311            NaN           NaN            NaN         NaN         NaN  
6312            NaN           NaN            NaN         NaN         NaN  
6313            NaN           NaN            NaN         NaN         NaN  
...             ...           ...            ...         ...         ...  
8698            NaN           NaN            NaN         NaN         NaN  
8699            NaN           NaN            NaN         NaN         NaN  
8700            NaN           NaN            NaN         NaN         NaN  
8701            NaN           NaN            NaN         NaN         NaN  
25722           NaN           NaN            NaN         NaN         NaN  

[2394 rows x 24 columns]

    '' Now that it's confirmed that apart from the "case number" column, all the other values are NaN, we can drop them.''

In [3096]:
sharks_ = sharks_[(sharks_["Case Number"] != "0") & (sharks_["Case Number"] != "xx")]
sharks_.sample(10)

Case Number         Date    Year        Type           Country  \
385     2015.07.25  25-Jul-2015  2015.0  Unprovoked         AUSTRALIA   
194   2016.12.11.b  11-Dec-2016  2016.0  Unprovoked  FRENCH POLYNESIA   
4984    1932.08.06  06-Aug-1932  1932.0    Provoked           BAHAMAS   
4005  1960.04.00.b     Apr-1960  1960.0    Provoked       PHILIPPINES   
1214    2008.08.28  28-Aug-2008  2008.0  Unprovoked               USA   
3979    1960.07.05  05-Jul-1960  1960.0  Unprovoked               USA   
6217       ND.0097      No date     0.0  Unprovoked               USA   
2399    1995.01.24  24-Jan-1995  1995.0  Unprovoked      SOUTH AFRICA   
2604    1991.04.03  03-Apr-1991  1991.0  Unprovoked               USA   
1177    2009.01.00     Jan-2009  2009.0  Unprovoked              CUBA   

                         Area                               Location  \
385                  Tasmania         Lachan Island, Mercury Passage   
194                  Tuamotos                         Takapoto Atoll   
4984  New Providence District          Nassau, New Providence Island   
4005             Luzon Island                                 Bataan   
1214                  Florida       New Smyrna Beach, Volusia County   
3979              Mississippi                       Mississippi City   
6217                  Florida                Key West, Monroe County   
2399            KwaZulu-Natal                               Isipingo   
2604                   Hawaii  One'ula Beach Park, 'Ewa Beach, O'ahu   
1177      Guantanamo Province                             Guantanamo   

                                               Activity  \
385                            Scallop diving on hookah   
194                                             Fishing   
4984  On expedition filming a feature movie & standi...   
4005  Fishing, hooked shark towed boat out to sea, s...   
1214                                            Surfing   
3979                  Pulling raft out to ride to shore   
6217                                        Kitesurfing   
2399                                           Swimming   
2604                               Sitting on surfboard   
1177                                       Spearfishing   

                               Name Sex   ...                        Species   \
385                  Damien Johnson    M  ...       White shark, 3.9 to 4.2 m   
194                            male    M  ...                             NaN   
4984              George Vanderbilt    M  ...                             NaN   
4005  Nicaso Balanoba & Julian Dona    M  ...                             NaN   
1214                    Thomas Gold    M  ...                             NaN   
3979              Henry Hanson, Jr.    M  ...                             NaN   
6217                     Paul Menta    M  ...                             NaN   
2399      Mthokozisi Cedrick Mpanza    M  ...        Tiger shark, 1.8 m [6']    
2604                  Todd R. Wenke    M  ...  "Shark had a very large girth"   
1177                           John  NaN  ...                      Bull shark   

                           Investigator or Source  \
385                                C. Black, GSAF   
194                       Tahiti Info, 12/12/2016   
4984                    New York Times, 8/7/1932    
4005     P. Gilbert (1961); H.D. Baldridge, p.152   
1214                           S. Petersohn, GSAF   
3979                 V.M. Coppleson (1962), p.249   
6217                                     Internet   
2399                             A. Gifford, GSAF   
2604  J. Borg, p.79; L. Taylor (1993), pp.110-111   
1177                           C. Johansson, GSAF   

                                 pdf  \
385           2015.07.25-Johnson.pdf   
194        2015.12.11.b-Takapoto.pdf   
4984       1932.08.06-Vanderbilt.pdf   
4005  1960.04.00.b-Balanoba-Dona.pdf   
1214             2008.08.28-Gold.pdf   
3979           1960.07.05-Hanson.pdf   
6217

    '''After deleting the case numbers with "0" and "xx" I want to check again that there are no rows with case number values but all the rest NaN. If we run the same code again, it should return 0 rows.'''

In [3097]:
nan_r = sharks_[sharks_.drop("Case Number", axis = 1).isna().all(axis= 1)]
nan_r

Empty DataFrame
Columns: [Case Number, Date, Year, Type, Country, Area, Location, Activity, Name, Sex , Age, Injury, Fatal (Y/N), Time, Species , Investigator or Source, pdf, href formula, href, Case Number.1, Case Number.2, original order, Unnamed: 22, Unnamed: 23]
Index: []

[0 rows x 24 columns]

    - Columns

In [3098]:
sharks_ = sharks_.dropna(axis=1, how='all')
sharks_.sample(10)

Case Number                  Date    Year          Type  \
1656    2004.07.15          15-Jul-2004   2004.0    Unprovoked   
5382    1911.10.25           25-Oct-1911  1911.0    Unprovoked   
1288  2008.02.21.R  Reported 21-Feb-2008  2008.0    Unprovoked   
3957  1960.10.00.d              Oct-1960  1960.0  Sea Disaster   
1164  2009.01.24.c           24-Jan-2009  2009.0          Boat   
58    2017.11.30.a           30-Nov-2017  2017.0    Unprovoked   
1051    2010.01.12           12-Jan-2010  2010.0    Unprovoked   
287     2016.04.13           13-Apr-2016  2016.0    Unprovoked   
5782  1883.10.20.R  Reported 20-Oct-1883  1883.0    Unprovoked   
2897    1984.03.17           17-Mar-1984  1984.0       Invalid   

                     Country                               Area  \
1656                   JAPAN                Wakayama Prefecture   
5382               AUSTRALIA                         Queensland   
1288        FRENCH POLYNESIA                    Society Islands   
3957  RED SEA / INDIAN OCEAN  Enroute from Suez to Aden (Yemen)   
1164             NEW ZEALAND                       North Island   
58                COSTA RICA                       Cocos Island   
1051            SOUTH AFRICA                   Western Province   
287                      USA                            Florida   
5782                 URUGUAY                    Maldonado coast   
2897                 SOMALIA                                NaN   

                                  Location  \
1656                                Susami   
5382                    Great Barrier Reef   
1288                                Tahiti   
3957                                   NaN   
1164                      Alderman Islands   
58                               Manuelita   
1051                             Fish Hoek   
287   Off Singer Island, Palm Beach County   
5782                                   NaN   
2897                                   NaN   

                                               Activity  \
1656  Fishing for squid aboard the trawler Shikishim...   
5382                                             Diving   
1288                                       Spearfishing   
3957  Cargo ship El Gamil entroute Suez to Yemen (Ad...   
1164                                            Fishing   
58                                         Scuba diving   
1051                                           Standing   
287                                        Spearfishing   
5782                           Fell overboard & rescuer   
2897                                             Murder   

                                     Name Sex   ...  \
1656                          Yuji Torimi    M  ...   
5382                                 Toby    M  ...   
1288                           Apia Hauta    M  ...   
3957                                  NaN    M  ...   
1164  7.2 m boat. Occupant Kelvin Travers  NaN  ...   
58                        Rohina Bhandari    F  ...   
1051                        Lloyd Skinner    M  ...   
287                        Kyle Senkowicz    M  ...   
5782              crew of the Galicia x 2    M  ...   
2897                           11 stoways    M  ...   

                                               Species   \
1656  Longfin mako shark, 3.5 m [11.5'], 350-kg [772...   
5382                                                NaN   
1288                                                NaN   
3957                                                NaN   
1164                                                NaN   
58                                  Tiger shark, female   
1051                                        White shark   
287                                      Bull shark, 7'   
5782                                                NaN   
2897  Shark involvement prior to death was not confi...   

                   Investigator or Source                              pdf  \
1656         Mainichi Shimbun, 7/17/2004           

    ''' Because it looked like Unamed 23 and 23 where full of Nans but didn't get deleted, I wanted to reassure thath the columns had at least some not null values'''

In [3099]:
sharks_[sharks_["Unnamed: 22"].notnull()]

Case Number         Date    Year        Type Country    Area  \
1478  2006.05.27  27-May-2006  2006.0  Unprovoked     USA  Hawaii   

                Location Activity          Name Sex   ... Species   \
1478  North Shore, O'ahu  Surfing  Bret Desmond    M  ...      NaN   

     Investigator or Source                     pdf  \
1478             R. Collier  2006.05.27-Desmond.pdf   

                                           href formula  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...    2006.05.27   

     Case Number.2 original order   Unnamed: 22 Unnamed: 23  
1478    2006.05.27         4825.0  stopped here         NaN  

[1 rows x 24 columns]

In [3100]:
sharks_[sharks_["Unnamed: 23"].notnull()]

Case Number                  Date    Year        Type  \
4415    1952.03.30           30-Mar-1952  1952.0  Unprovoked   
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0    Provoked   

                   Country         Area                    Location  \
4415  NETHERLANDS ANTILLES      Curacao                         NaN   
5840                   USA  Connecticut  Branford, New Haven County   

                                             Activity              Name Sex   \
4415  Went to aid of child being menaced by the shark       A.J. Eggink    M   
5840                                          Fishing  Captain Pattison    M   

      ...                                           Species   \
4415  ...  Bull shark, 2.7 m [9'] was captured & dragged ...   
5840  ...                                                NaN   

                                 Investigator or Source  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   
5840                       St. Joseph Herald, 9/14/1878   

                            pdf  \
4415      1952.03.30-Eggink.pdf   
5840  1878.09.14.R-Pattison.pdf   

                                           href formula  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...  1878.09.14.R   

     Case Number.2 original order Unnamed: 22      Unnamed: 23  
4415    1952.03.30         1888.0         NaN           Teramo  
5840  1878.09.14.R          463.0         NaN  change filename  

[2 rows x 24 columns]

    3. Check for repeated case ID's so we don't have repeated information = false data.

    - First I'm going to check the names of the columns and delete or replace the white spaces

In [3101]:
sharks_.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [3102]:
sharks_.rename(columns={"Sex ":"Sex"}, inplace= True)
sharks_.rename(columns={"Species ":"Species"}, inplace= True)
sharks_

Case Number         Date    Year        Type             Country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   Area                             Location      Activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                                  NaN  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300                NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       Name Sex  ...          Species  \
0               Julie Wolfe   F  ...      White shark   
1           Adyson McNeely    F  ...              NaN   
2               John Denges   M  ...              NaN   
3                      male   M  ...        2 m shark   
4            Gustavo Ramos    M  ...  Tiger shark, 3m   
...                     ...  ..  ...              ...   
6297                   male   M  ...              NaN   
6298                  Ahmun   M  ...              NaN   
6299  Coast Guard personnel   M  ...              NaN   
6300        Jules Patterson   M  ...              NaN   
6301                   male   M  ...              NaN   

                     Investigator or Source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://sharkattackfi

In [3168]:
comp_dict = {i : i.lower().replace(" ", "_") for i in sharks_.columns}
comp_dict
'''Changing the names in the columns'''

'Changing the names if the columns'

In [3104]:
sharks_.rename(columns = comp_dict, inplace=True)
sharks_

case_number         date    year        type             country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   area                             location      activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                                  NaN  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300                NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       name sex  ...          species  \
0               Julie Wolfe   F  ...      White shark   
1           Adyson McNeely    F  ...              NaN   
2               John Denges   M  ...              NaN   
3                      male   M  ...        2 m shark   
4            Gustavo Ramos    M  ...  Tiger shark, 3m   
...                     ...  ..  ...              ...   
6297                   male   M  ...              NaN   
6298                  Ahmun   M  ...              NaN   
6299  Coast Guard personnel   M  ...              NaN   
6300        Jules Patterson   M  ...              NaN   
6301                   male   M  ...              NaN   

                     investigator_or_source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href_formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://sharkattackfi

    - Now that the names are changed I can look for duplicates

    ''' I look for duplicates in the case_number column because it's supposed to be unique.'''

In [3105]:
sharks_.case_number.duplicated().sum()

16

    ''' Having 16 duplicates I wanted to check if the only duplicated value was the case number or in fact all the row, before droping them.'''

In [3106]:
cs = sharks_["case_number"]

In [3107]:
sharks_[cs.isin(cs[cs.duplicated()])].sort_values("case_number")

case_number                  date    year        type       country  \
5444    1907.10.16.R  Reported 16-Oct-1907  1907.0  Unprovoked         CHINA   
5443    1907.10.16.R  Reported 16-Oct-1907  1907.0  Unprovoked         CHINA   
5355    1913.08.27.R  Reported 27-Aug-1913  1913.0     Invalid           USA   
5354    1913.08.27.R  Reported 27-Aug-1913  1913.0     Invalid           USA   
5323  1915.07.06.a.R  Reported 06-Jul-1915  1915.0  Unprovoked        MEXICO   
5322  1915.07.06.a.R  Reported 06-Jul-1915  1915.0  Unprovoked        MEXICO   
5261    1920.00.00.b                  1920  1920.0  Unprovoked   NEW ZEALAND   
5260    1920.00.00.b                 1920s  1920.0  Unprovoked       JAMAICA   
5214    1923.00.00.a                  1923  1922.0    Provoked           USA   
5213    1923.00.00.a                  1923  1923.0    Provoked           USA   
3815    1962.06.11.b           11-Jun-1962  1962.0  Unprovoked           USA   
3814    1962.06.11.b           11-Jun-1962  1962.0  Unprovoked           USA   
3544      1966.12.26           26-Dec-1966  1966.0  Unprovoked     AUSTRALIA   
3543      1966.12.26           26-Dec-1966  1966.0  Unprovoked     SINGAPORE   
3063      1980.07.00        Early Jul-1980  1980.0  Unprovoked           USA   
3062      1980.07.00         Late Jul-1980  1980.0  Unprovoked           USA   
2938      1983.06.15           15-Jun-1983  1983.0     Invalid         ITALY   
2937      1983.06.15           15-Jun-1983  1983.0  Unprovoked       BAHAMAS   
2635      1990.05.10           10-May-1990  1990.0  Unprovoked     AUSTRALIA   
2634      1990.05.10           10-May-1990  1990.0  Unprovoked     AUSTRALIA   
1600      2005.04.06           06-Apr-2005  2005.0     Invalid      HONDURAS   
1599      2005.04.06           06-Apr-2005  2005.0  Unprovoked           USA   
1437      2006.09.02           02-Sep-2006  2006.0  Unprovoked  SOUTH AFRICA   
1436      2006.09.02           02-Sep-2006  2006.0  Unprovoked           USA   
1064      2009.12.18           18-Dec-2009  2009.0     Invalid  SOUTH AFRICA   
1063      2009.12.18           18-Dec-2009  2009.0  Unprovoked  SOUTH AFRICA   
747     2012.09.02.b           02-Sep-2012  2012.0    Provoked           USA   
746     2012.09.02.b           02-Sep-2012  2012.0  Unprovoked           USA   
616       2013.10.05           10-Oct-2013  2013.0  Unprovoked           USA   
615       2013.10.05           06-Oct-2013  2013.0  Unprovoked           USA   
523       2014.08.02           02-Aug-2014  2014.0  Unprovoked           USA   
522       2014.08.02           02-Aug-2014  2014.0  Unprovoked           USA   

                       area  \
5444              Hong Kong   
5443              Hong Kong   
5355             New Jersey   
5354             New Jersey   
5323                    NaN   
5322             Tamaulipas   
5261           North Island   
5260    Westmoreland Parish   
5214             New Jersey   
5213             New Jersey   
3815             California   
3814             California   
3544        New South Wales   
3543                    NaN   
3063                Florida   
3062         North Carolina   
2938        Northwest Italy   
2937                    NaN   
2635             Queensland   
2634             Queensland   
1600            Bay Islands   
1599                Florida   
1437  Western Cape Province   
1436                Florida   
1064          KwaZulu-Natal   
1063  Eastern Cape Province   
747                  Hawaii   
746                 Florida   
616                 Florida   
615              California   
523                 Florida   
522                 Florida   

                                               location  \
5444    Sharp Peak, Sai Kung Peninsula, New Territories   
5443    Sharp Peak, Sai Kung Peninsula, New Territories   
5355                       Spring Lake, Monmouth County   
5354                            Lavalette, Ocean County   
5323                                    Santa Mar

    ''' Now that it's confirmed that not all the 16 raws are fully duplicated I can look for fully duplicated rows and drop those. Also if needed we'll need to change the names of the duplicated case_numbers'''

In [3108]:
sharks_=sharks_.drop_duplicates()
sharks_

case_number         date    year        type             country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   area                             location      activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                                  NaN  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300                NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       name sex  ...          species  \
0               Julie Wolfe   F  ...      White shark   
1           Adyson McNeely    F  ...              NaN   
2               John Denges   M  ...              NaN   
3                      male   M  ...        2 m shark   
4            Gustavo Ramos    M  ...  Tiger shark, 3m   
...                     ...  ..  ...              ...   
6297                   male   M  ...              NaN   
6298                  Ahmun   M  ...              NaN   
6299  Coast Guard personnel   M  ...              NaN   
6300        Jules Patterson   M  ...              NaN   
6301                   male   M  ...              NaN   

                     investigator_or_source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href_formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://sharkattackfi

    4. Now I want to check the columns that we have to do a more in depth cleaning or at least to decide how to proceede.

In [3109]:
sharks_.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

In [3110]:
sharks_["href"].sample(5)

2201    http://sharkattackfile.net/spreadsheets/pdf_di...
5544    http://sharkattackfile.net/spreadsheets/pdf_di...
6299    http://sharkattackfile.net/spreadsheets/pdf_di...
5052    http://sharkattackfile.net/spreadsheets/pdf_di...
2554    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href, dtype: object

In [3111]:
sharks_["href_formula"].sample(5)

2480    http://sharkattackfile.net/spreadsheets/pdf_di...
785     http://sharkattackfile.net/spreadsheets/pdf_di...
2319    http://sharkattackfile.net/spreadsheets/pdf_di...
2259    http://sharkattackfile.net/spreadsheets/pdf_di...
3397    http://sharkattackfile.net/spreadsheets/pdf_di...
Name: href_formula, dtype: object

In [3112]:
sharks_["original_order"].sample(5)

2819    3484.0
4392    1911.0
693     5610.0
6183     120.0
1548    4755.0
Name: original_order, dtype: float64

In [3113]:
sharks_["original_order"].sort_values()

6301       2.0
6300       3.0
6299       4.0
6298       5.0
6297       6.0
         ...  
4       6299.0
3       6300.0
2       6301.0
1       6302.0
0       6303.0
Name: original_order, Length: 6302, dtype: float64

    '''Seeing that we can assume that original order it's kind of an index so seeing as there are some case_number values repeated we can check for duplicates in original order and maybe filter from there.'''

In [3114]:
sharks_.original_order.duplicated().sum()

1

In [3115]:
oo = sharks_["original_order"]

In [3116]:
sharks_[oo.isin(oo[oo.duplicated()])].sort_values("original_order")

case_number         date    year        type       country  \
5733  1888.01.22  22-Jan-1888  1888.0     Boating     AUSTRALIA   
5734  1888.02.00     Feb-1888  1888.0  Unprovoked  SOUTH AFRICA   

                       area              location                  activity  \
5733        New South Wales         Sydney Harbor                    Rowing   
5734  Eastern Cape Province  Mzimvubu River mouth  Crossing the river mouth   

       name sex  ... species  \
5733  Burke   M  ...     NaN   
5734   male   M  ...     NaN   

                                 investigator_or_source  \
5733                                    Star, 1/23/1888   
5734  Cape Mercantile Advertiser, 2/15/1888, M. Levi...   

                          pdf  \
5733     1888.01.22-Burke.pdf   
5734  1888.02.00-Mzimvubu.pdf   

                                           href_formula  \
5733  http://sharkattackfile.net/spreadsheets/pdf_di...   
5734  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
5733  http://sharkattackfile.net/spreadsheets/pdf_di...    1888.01.22   
5734  http://sharkattackfile.net/spreadsheets/pdf_di...    1888.02.00   

     case_number.2 original_order unnamed:_22 unnamed:_23  
5733    1888.01.22          569.0         NaN         NaN  
5734    1888.02.00          569.0         NaN         NaN  

[2 rows x 24 columns]

    ''' There's only one duplicated value but the rest of the column values are different so seeing that the original order starts with 2, what I can do is change one of the values for 1, so that it gets more organized.Also, that mens that the drop duplicates done before didn't work because this value was always different, so I'm going to add also a drop duplicates excluding the original order column.'''

In [3117]:
sharks_.loc[5733]

case_number                                                      1888.01.22
date                                                            22-Jan-1888
year                                                                 1888.0
type                                                                Boating
country                                                           AUSTRALIA
area                                                        New South Wales
location                                                      Sydney Harbor
activity                                                             Rowing
name                                                                  Burke
sex                                                                       M
age                                                                     NaN
injury                    Shark bit boat, but no injury to occupant who ...
fatal_(y/n)                                                               N
time        

In [3118]:
sharks_.at[5733, "original_order"] = 1

In [3119]:
sharks_["original_order"].sort_values()

5733       1.0
6301       2.0
6300       3.0
6299       4.0
6298       5.0
         ...  
4       6299.0
3       6300.0
2       6301.0
1       6302.0
0       6303.0
Name: original_order, Length: 6302, dtype: float64

In [3120]:
sharks_.original_order.duplicated().sum()

0

In [3121]:
sharks_.drop_duplicates(subset=sharks_.columns.difference(["original_order"]))

case_number         date    year        type             country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   area                             location      activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                                  NaN  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300                NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       name sex  ...          species  \
0               Julie Wolfe   F  ...      White shark   
1           Adyson McNeely    F  ...              NaN   
2               John Denges   M  ...              NaN   
3                      male   M  ...        2 m shark   
4            Gustavo Ramos    M  ...  Tiger shark, 3m   
...                     ...  ..  ...              ...   
6297                   male   M  ...              NaN   
6298                  Ahmun   M  ...              NaN   
6299  Coast Guard personnel   M  ...              NaN   
6300        Jules Patterson   M  ...              NaN   
6301                   male   M  ...              NaN   

                     investigator_or_source                             pdf  \
0                          R. Collier, GSAF            2018.06.25-Wolfe.pdf   
1            K.McMurray, TrackingSharks.com          2018.06.18-McNeely.pdf   
2            K.McMurray, TrackingSharks.com           2018.06.09-Denges.pdf   
3                            B. Myatt, GSAF        2018.06.08-Arrawarra.pdf   
4                                 A .Kipper            2018.06.04-Ramos.pdf   
...                                     ...                             ...   
6297       H. Taunton; N. Bartlett,  p. 234          ND-0005-RoebuckBay.pdf   
6298  H. Taunton; N. Bartlett,  pp. 233-234               ND-0004-Ahmun.pdf   
6299   F. Schwartz, p.23; C. Creswell, GSAF  ND-0003-Ocracoke_1900-1905.pdf   
6300                    The Sun, 10/20/1938      ND-0002-JulesPatterson.pdf   
6301                             S.W. Baker              ND-0001-Ceylon.pdf   

                                           href_formula  \
0     http://sharkattackfile.net/spreadsheets/pdf_di...   
1     http://sharkattackfile.net/spreadsheets/pdf_di...   
2     http://sharkattackfile.net/spreadsheets/pdf_di...   
3     http://sharkattackfile.net/spreadsheets/pdf_di...   
4     http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                 ...   
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://sharkattackfi

    ''' Now that we see that there were no duplicates to start with as even excluding the original order column to drop duplicates the row count keeps on being 6302; I can keep on looking at some more columns before choosing which ones to dellete and which ones to keep. I want to chek from href and href formula which one has more values because they are the same and so I'll just keep the one with more info.'''

In [3122]:
sharks_["href_formula"].notna().sum()

6301

In [3123]:
sharks_["href"].notna().sum()

6302

    5. Now I'm going to drop the columns that don't give me any usefull information or repeated information. 
        I'll be droping:
        - href and href_formula
        - case number 1 and 2 because we already got the case number
        - also the columns unnamed 22 and 23 because for 1/2 values of unusefull info I don't want ot keed them. 
        - pdf I'll also drop because I won't use it.
        - And finally the original order, I realized it serves me no purpose.


    '"For future projects maybe I should start by doing this just after droping the nan rows and columns and before the duplicates."'

In [3124]:
sharks_ = sharks_.drop(columns=["href", "href_formula", "case_number.1", "case_number.2", "unnamed:_22" , "unnamed:_23", "original_order", "pdf"])
sharks_.sample()

case_number         date    year     type    country             area  \
5019  1931.05.02  02-May-1931  1931.0  Boating  AUSTRALIA  New South Wales   

     location activity                                             name  sex  \
5019  Kempsey      NaN  pilot boat, occupants; Captain McAlister & crew  NaN   

      age                                       injury fatal_(y/n) time  \
5019  NaN  No injury to occupants; oar & rudder bitten           N  NaN   

                   species                           investigator_or_source  
5019  "a school of sharks"  Sydney Morning Herald, 6/5/1931; Whitley, p.262

    6. I'm going to do a more specific cleaning for the resulting columns focusing on the ones I want to work with. 

    ''' I'm going to start "sex", then follow with "age", then "Fatal Y/N" anf then "Activity"'.

    - Sex:

In [3125]:
sharks_["sex"].value_counts()

sex
M      5094
F       637
M         2
N         2
lli       1
.         1
Name: count, dtype: int64

In [3126]:
sharks_["sex"].unique() #To check how it's spelled.

array(['F', 'M', nan, 'M ', 'lli', 'N', '.'], dtype=object)

    ''' To clean the sex column I'm going to do a function to apply it there."'

In [3127]:
def sex_clean(x):
    if x != "M" and x != "F" and x != "M ":
        return "unknown"
    elif x == "M ":
        return "M"
    else:
        return x
    
    sharks_["sex"] = sharks_["sex"].apply(sex_clean)

    return vvhgcghf


In [3128]:
sharks_["sex"] = sharks_["sex"].apply(sex_clean)

In [3129]:
sharks_["sex"].unique()

array(['F', 'M', 'unknown'], dtype=object)

    - Fatal Y/N:

In [3130]:
sharks_["fatal_(y/n)"].value_counts()

fatal_(y/n)
N          4293
Y          1388
UNKNOWN      71
 N            7
M             1
2017          1
N             1
y             1
Name: count, dtype: int64

In [3131]:
sharks_["fatal_(y/n)"].unique()

array(['N', 'Y', nan, 'M', 'UNKNOWN', '2017', ' N', 'N ', 'y'],
      dtype=object)

In [3132]:
def fatal_clean(x):
    if x != "N" and x != "Y" and x != "N " and x != " N" and x != "y":
        return "unknown"
    elif x == "N " or x == " N":
        return "N"
    elif x == "y":
        return "Y"
    else:
        return x

In [3133]:
sharks_["fatal_(y/n)"] = sharks_["fatal_(y/n)"].apply(fatal_clean)

In [3134]:
sharks_["fatal_(y/n)"].unique()

array(['N', 'Y', 'unknown'], dtype=object)

    - Type of attack

In [3135]:
sharks_["type"].value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Sea Disaster     239
Boating          203
Boat             137
Questionable       2
Boatomg            1
Name: count, dtype: int64

In [3136]:
sharks_["type"].unique() 

array(['Boating', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

In [3137]:
pattern = r'\w*[Bb]oat\w*'
sharks_["type"] = sharks_['type'].str.replace(pattern, 'Boat', regex=True)

''' Regex code to unify boat and boatomg and boating'''

' Regex code to unify boat and boatomg and boating'

In [3138]:
sharks_["type"].value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Boat             341
Sea Disaster     239
Questionable       2
Name: count, dtype: int64

In [3167]:
def clean_type(i):
    if pd.isna(i):
        return "unknown"
    else: 
        return i
    
''' I created this function because the data was basically clean but I wanted to have objects
 and nans are floats so i wanted to transform the nan to "unknown" . I could have done it with a for loop 
 but I think that one function is actually functional :')'''

' I created this function because the data was basically clean but I wanted to have objects\n and nans are floats so i wanted to transform the nan to "unknown" . I could have done it with a for loop \n but I think that one function is actually functional :\')'

In [3140]:
sharks_["type"] = sharks_["type"].apply(clean_type)

In [3141]:
sharks_["type"].unique() 

array(['Boat', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster', 'unknown'], dtype=object)

    - Year

In [3142]:
sharks_["year"].unique()

array([2018., 2017.,   nan, 2016., 2015., 2014., 2013., 2012., 2011.,
       2010., 2009., 2008., 2007., 2006., 2005., 2004., 2003., 2002.,
       2001., 2000., 1999., 1998., 1997., 1996., 1995., 1984., 1994.,
       1993., 1992., 1991., 1990., 1989., 1969., 1988., 1987., 1986.,
       1985., 1983., 1982., 1981., 1980., 1979., 1978., 1977., 1976.,
       1975., 1974., 1973., 1972., 1971., 1970., 1968., 1967., 1966.,
       1965., 1964., 1963., 1962., 1961., 1960., 1959., 1958., 1957.,
       1956., 1955., 1954., 1953., 1952., 1951., 1950., 1949., 1948.,
       1848., 1947., 1946., 1945., 1944., 1943., 1942., 1941., 1940.,
       1939., 1938., 1937., 1936., 1935., 1934., 1933., 1932., 1931.,
       1930., 1929., 1928., 1927., 1926., 1925., 1924., 1923., 1922.,
       1921., 1920., 1919., 1918., 1917., 1916., 1915., 1914., 1913.,
       1912., 1911., 1910., 1909., 1908., 1907., 1906., 1905., 1904.,
       1903., 1902., 1901., 1900., 1899., 1898., 1897., 1896., 1895.,
       1894., 1893.,

In [3166]:

def clean_year(year):
    pattern_year = r'\d{4}'
    str_pattern = r'\d{4}\.'
    if re.match(str_pattern, str(year)):
        str_ = re.match(pattern_year, str(year)).group()
        return int(str_)
    else:
        return 'unknown'
    
'''To make it 4 digits and if not unknown and to turn them into ints'''


'To make it 4 digits and if not unknown and to turn them into ints'

In [3144]:
sharks_["year"] = sharks_["year"].apply(clean_year)

In [3145]:
sharks_["year"].unique()

array([2018, 2017, 'unknown', 2016, 2015, 2014, 2013, 2012, 2011, 2010,
       2009, 2008, 2007, 2006, 2005, 2004, 2003, 2002, 2001, 2000, 1999,
       1998, 1997, 1996, 1995, 1984, 1994, 1993, 1992, 1991, 1990, 1989,
       1969, 1988, 1987, 1986, 1985, 1983, 1982, 1981, 1980, 1979, 1978,
       1977, 1976, 1975, 1974, 1973, 1972, 1971, 1970, 1968, 1967, 1966,
       1965, 1964, 1963, 1962, 1961, 1960, 1959, 1958, 1957, 1956, 1955,
       1954, 1953, 1952, 1951, 1950, 1949, 1948, 1848, 1947, 1946, 1945,
       1944, 1943, 1942, 1941, 1940, 1939, 1938, 1937, 1936, 1935, 1934,
       1933, 1932, 1931, 1930, 1929, 1928, 1927, 1926, 1925, 1924, 1923,
       1922, 1921, 1920, 1919, 1918, 1917, 1916, 1915, 1914, 1913, 1912,
       1911, 1910, 1909, 1908, 1907, 1906, 1905, 1904, 1903, 1902, 1901,
       1900, 1899, 1898, 1897, 1896, 1895, 1894, 1893, 1892, 1891, 1890,
       1889, 1888, 1887, 1886, 1885, 1884, 1883, 1882, 1881, 1880, 1879,
       1878, 1877, 1876, 1875, 1874, 1873, 1872, 187

In [3146]:
sharks_.dtypes

case_number               object
date                      object
year                      object
type                      object
country                   object
area                      object
location                  object
activity                  object
name                      object
sex                       object
age                       object
injury                    object
fatal_(y/n)               object
time                      object
species                   object
investigator_or_source    object
dtype: object

In [3165]:
for i in sharks_["year"]:
    if i != "unknown":
        print(type(i))

''' To be sure that I transformed the years into int'''

<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class 'int'>
<class

' To be sure that I transformed the years into int'

    - Age:  

In [3148]:
sharks_["age"].unique()

array(['57', '11', '48', nan, '18', '52', '15', '12', '32', '10', '21',
       '34', '30', '60', '33', '29', '54', '41', '37', '56', '19', '25',
       '69', '38', '55', '35', '46', '45', '14', '40s', '28', '20', '24',
       '26', '49', '22', '7', '31', '17', '40', '13', '42', '3', '8',
       '50', '16', '82', '73', '20s', '68', '51', '39', '58', 'Teen',
       '47', '61', '65', '36', '66', '43', '60s', '9', '72', '59', '6',
       '27', '64', '23', '71', '44', '62', '63', '70', '18 months', '53',
       '30s', '50s', 'teen', '77', '74', '28 & 26', '5', '86', '18 or 20',
       '12 or 13', '46 & 34', '28, 23 & 30', 'Teens', '36 & 26',
       '8 or 10', '84', '\xa0 ', ' ', '30 or 36', '6½', '21 & ?', '75',
       '33 or 37', 'mid-30s', '23 & 20', ' 30', '7      &    31', ' 28',
       '20?', "60's", '32 & 30', '16 to 18', '87', '67', 'Elderly',
       'mid-20s', 'Ca. 33', '74 ', '45 ', '21 or 26', '20 ', '>50',
       '18 to 22', 'adult', '9 & 12', '? & 19', '9 months', '25 to 35',
  

In [3149]:
def extract_age(x):

    '''first I transform manually the nan and the values with only 1 to 3 variations
      like teens, adult, 18 month.'''
    
    if pd.isna(x):
        return "unknown"
    if x == "teen" or x == "Teen" or x == "Teens":
        return "15"
    if x == "adult" or x == "(adult)" or x == "middle-aged":
        return "50"
    if x == "18 months":
        return "2"
    
    '''create a pattern to make the averege of the different ages of the values that give you 
    an age, then & ,or, to. Then applied the pattern, transformed the strings into ints to 
    get the average, get the items out of the list and transform them into string again.'''

    age_av = re.findall(r'(\d{1,2})\s*(&|or|to)\s*(\d{1,2})', str(x))
    
    if age_av:
        average_ages = [round((int(match[0]) + int(match[2])) / 2) for match in age_av]
        return str(average_ages[0])
    
    '''Also made a pattern to fins the values with text and numbers to catch 
    the 1 or 2 numbers to make the age.'''

    age_match = re.search(r'\d{1,2}', str(x))
    if age_match:
        return age_match.group()
    return "unknown"


In [3150]:
sharks_["age"] = sharks_["age"].apply(extract_age)

In [3151]:
sharks_["age"].unique()

array(['57', '11', '48', 'unknown', '18', '52', '15', '12', '32', '10',
       '21', '34', '30', '60', '33', '29', '54', '41', '37', '56', '19',
       '25', '69', '38', '55', '35', '46', '45', '14', '40', '28', '20',
       '24', '26', '49', '22', '7', '31', '17', '13', '42', '3', '8',
       '50', '16', '82', '73', '68', '51', '39', '58', '47', '61', '65',
       '36', '66', '43', '9', '72', '59', '6', '27', '64', '23', '71',
       '44', '62', '63', '70', '2', '53', '77', '74', '5', '86', '84',
       '75', '87', '67', '1', '81', '78'], dtype=object)

    - Activity:

In [3152]:
sharks_["activity"].unique()

array(['Paddling', 'Standing', 'Surfing', ...,
       'Crew swimming alongside their anchored ship',
       '4 men were bathing', 'Wreck of  large double sailing canoe'],
      dtype=object)

In [3153]:
sharks_["activity"].isna().sum()

544

In [3154]:
sharks_["activity"].notna().sum()

5758

In [3155]:
sharks_["activity"].value_counts()

activity
Surfing                                   971
Swimming                                  869
Fishing                                   431
Spearfishing                              333
Bathing                                   162
                                         ... 
Playing with a frisbee in the shallows      1
Sinking of the ferryboat Dumaguete          1
Wreck of the Storm King                     1
Feeding mullet to sharks                    1
Wreck of  large double sailing canoe        1
Name: count, Length: 1532, dtype: int64

In [3156]:
set(sharks_.activity)

{'Sea disaster',
 'Sea disaster, foundering of the cargo vessle M/V Dorolonda',
 'During "an exhibition" he was tied in sack & thrown overboard ',
 'Scuba diving / culling lionfish',
 'Put foot inside mouth of supposedly dead shark',
 'Fishing from dinghy',
 '"Climbing up to ship after repairing the stern in water"',
 'Bather',
 'Windsurfing, but sitting on his board',
 'Fishingat',
 "Accidentally stood on hooked shark's tail before attempting to gut it ",
 'Swimming, after single-engine aircraft went down in the sea',
 'Fishing for bottom fish',
 'Kayaking or Wave skiing',
 'Abandoning burning ship Captain George in raging seas',
 'Shark hoisted on board liner Tahiti',
 'Diving for sea urchins',
 'Jumped overboard from Norwegian steamship Venator',
 'Standing',
 'Vehicle plunged over cliff into the water',
 'Fell off surf ski',
 'Unknown',
 'Fishing for squid aboard the trawler Shikishima-Maru when the shark leapt into the boat',
 'On fishing boat & trailing hand in the water',
 'Padd

    ''' As we can see there are tons of activities and explanations, so I'm going to create a function to count key words and from that create the redex patterns to clean the column.'''

In [3157]:
def find_key_words(items):
    items = [str(i) for i in items]
    text = ' '.join(items)
    words = text.split()
    counter = Counter(x for x in words if len(x)>2)
    rkw= [word for word, count in counter.items() if count > 1]
    return rkw


In [3158]:
activities = set(sharks_.activity)
repeated_key_words = find_key_words(activities)
for word in repeated_key_words:
    print(word)

Sea
disaster
disaster,
the
cargo
M/V
During
was
tied
thrown
overboard
Scuba
diving
Put
foot
inside
mouth
supposedly
dead
shark
Fishing
from
dinghy
ship
after
repairing
stern
but
sitting
his
board
Accidentally
stood
hooked
shark's
tail
before
attempting
Swimming,
aircraft
went
down
sea
for
bottom
fish
Kayaking
Wave
skiing
Abandoning
burning
Captain
George
seas
Shark
liner
Tahiti
Diving
urchins
Jumped
Norwegian
steamship
Standing
plunged
over
cliff
into
water
Fell
off
surf
ski
squid
aboard
trawler
when
leapt
boat
fishing
trailing
hand
Ship
anchor
man
working
its
Free
wading
back
shore
Dived
Playing
with
Hurricane
Gathering
shells
Clinging
overturned
rowing
Crossing
saw
shoal
fish,
dived
overboard,
had
speared
air
out
canoe
Adrift
sinking
motor
yacht
Disaster,
Swimming
between
anchored
pearling
Sitting
surfboard
feeding
Dragging
Spearing
ditched
torpedoed
sunk
propeller
Navy
The
troopship
German
raider
Spearfishing,
swimming
surface
Body
surfing
treading
Bathing
feet
river
trochus
sank
du

    ''' Now that we have some key words we can put those into lists to create a function to find those words in the column to replace for the full strings.'''

In [3159]:
boat_keywords = ['boat', 'boating', 'racing', 'barqued', 'sinking', 'ship', 'wreck', 'raft', 'cutter', 'bark', 'submarine', 'yacht', 'ferry']
fishing_keywords = ['fishing', 'fish', 'spearfishing', 'netting', 'wade-fishing', 'hunting', 'fishingat', 'shrimping', 'gigging', 'picking', 'net']
air_accidents_keywords = ['aircraft', 'airliner', 'constellation', 'parachute']
swimming_keywords = ['swimming', 'riding']
diving_keywords = ['diving', 'diver', 'photographing', 'dive', 'skindiving']
surf_keywords = ['surfing', 'surf', 'boogie', 'surf-skiing', 'paddeling', 'boarding', 'board', 'overboard', 'treading']
non_motor_keywords = ['kayak', 'canoa', 'canoe']
dangerous_keywords = ['explosives', 'anesthesize', 'investigating', 'pulling', 'filming', 'defecating', 'chumming', 'wrangling', 'murder', 'searching']
bathing_keywords = ['playing', 'bath', 'bathing', 'crouching', 'floating', 'standing', 'sitting', 'dangling']
natural_disaster_keywords = ['hurricane', 'natural']

    ''' I think there was another way using the lists, but I gave up. This is what I ended up doing:'''

In [3160]:
def activity_categorize(input_str):


    if pd.isna(input_str):
        return "other"    
    if re.search(r'.*swim.*', input_str, re.IGNORECASE):
        return "swimming"
    if re.search(r'.*fishing.*', input_str, re.IGNORECASE) or re.search(r'.*hunting.*', input_str, re.IGNORECASE) or re.search(r'.*net.*', input_str, re.IGNORECASE):
        return "fishing"
    if re.search(r'.*surf.*', input_str, re.IGNORECASE) or re.search(r'.*board.*', input_str, re.IGNORECASE) or re.search(r'.*paddeling.*', input_str, re.IGNORECASE) or re.search(r'.*treading.*', input_str, re.IGNORECASE):
        return "surfing"
    if re.search(r'.*div.*', input_str, re.IGNORECASE):
        return "diving"
    if re.search(r'.*boat.*', input_str, re.IGNORECASE) or re.search(r'.*sink.*', input_str, re.IGNORECASE) or re.search(r'.*ship.*', input_str, re.IGNORECASE) or re.search(r'.*bar.*', input_str, re.IGNORECASE) or re.search(r'.*wreck.*', input_str, re.IGNORECASE) or re.search(r'.*raft.*', input_str, re.IGNORECASE)or re.search(r'.*yacht.*', input_str, re.IGNORECASE)or re.search(r'.*submarine.*', input_str, re.IGNORECASE):
        return "sailing"
    if re.search(r'.*air.*', input_str, re.IGNORECASE) or re.search(r'.*constellation.*', input_str, re.IGNORECASE) or re.search(r'.*parachute.*', input_str, re.IGNORECASE):
        return "air accidents"
    if re.search(r'.*cano.*', input_str, re.IGNORECASE) or re.search(r'.*kayak.*', input_str, re.IGNORECASE):
        return "non motor"
    if re.search(r'.*explosive.*', input_str, re.IGNORECASE) or re.search(r'.*anestesize.*', input_str, re.IGNORECASE) or re.search(r'.*pull.*', input_str, re.IGNORECASE) or re.search(r'.*investigate.*', input_str, re.IGNORECASE) or re.search(r'.*film.*', input_str, re.IGNORECASE) or re.search(r'.*photo.*', input_str, re.IGNORECASE) or re.search(r'.*defecat.*', input_str, re.IGNORECASE) or re.search(r'.*chumm.*', input_str, re.IGNORECASE) or re.search(r'.*wrang.*', input_str, re.IGNORECASE) or re.search(r'.*murder.*', input_str, re.IGNORECASE) or re.search(r'.*search.*', input_str, re.IGNORECASE):
        return "shark interacting"
    if re.search(r'.*natural.*', input_str, re.IGNORECASE) or re.search(r'.*hurricane.*', input_str, re.IGNORECASE):
        return "natural disasters"
    if re.search(r'.*bath.*', input_str, re.IGNORECASE) or re.search(r'.*float.*', input_str, re.IGNORECASE) or re.search(r'.*stand.*', input_str, re.IGNORECASE) or re.search(r'.*sit.*', input_str, re.IGNORECASE) or re.search(r'.*dangling.*', input_str, re.IGNORECASE) or re.search(r'.*crouching.*', input_str, re.IGNORECASE):
        return "bathing"
    return "other"

''' Function to categorize the activities: 
    - First I descard the nan because they are floats and it breaks the function.
    - Then with regex I input the patterns on how to categorize
    ps: literally crying that it worked, spent like 6 hours on that.'''

' Function to categorize the activities: \n    - First I descard the nan because they are floats and it breaks the function.\n    - Then with regex I input the patterns on how to categorize\n    ps: literally crying that it worked, spent like 6 hours on that.'

In [3161]:
sharks_["activity"] = sharks_["activity"].apply(activity_categorize)

In [3162]:
sharks_["activity"].unique()

array(['other', 'bathing', 'surfing', 'diving', 'swimming', 'fishing',
       'sailing', 'non motor', 'shark interacting', 'air accidents',
       'natural disasters'], dtype=object)

In [3163]:
sharks_["activity"].value_counts()

activity
surfing              1521
other                1229
fishing              1202
swimming             1123
diving                522
bathing               360
sailing               229
non motor              59
shark interacting      33
air accidents          20
natural disasters       4
Name: count, dtype: int64

    7. Once done with the basic cleaning let's export the dataframe as csv and start analizing from there.

In [3164]:
sharks_.to_csv(f"../data/cleaned_attacks.csv", index = False)